### PR TITLE
[SQLDB ]Collation validation

### PR DESCRIPTION
**Pull request by siyangMicrosoft**
Adding collation validation in target database selection page.
1. collecting source database info including state, size, collation etc in the step 1 (select databases to assess and migrate)
2. comparing source database collation and target database collation in target selection page.
image

### CODE REVIEW

**Comment by siyangMicrosoft** [comment]
@microsoft-github-policy-service agree [company="Microsoft"]

Signature: Siyang Yao

Date: 1/05/2023
________________________________

**Comment by siyangMicrosoft** [comment]
@microsoft-github-policy-service agree company="Microsoft"
________________________________

File: extensions/sql-migration/src/constants/strings.ts [hunk 1]
```diff
@@ -410,6 +410,21 @@ export function SQL_TARGET_CONNECTION_SOURCE_NOT_MAPPED(sourceDatabaseName: stri
...
 
+//`Database mapping error.  Source database collation ({0}) is not same to target database collation ({1}). Source database: {2}; Target database: {3}.`
+export function SQL_TARGET_SOURCE_COLLATION_NOT_SAME(
+	sourceDatabaseName: string,
+	targetDatabaseName: string,
+	sourceDatabaseCollation: string,
+	targetDatabaseCollation: string): string {
+	return localize(
+		'sql.migration.wizard.target.source.collation.error',
+		"Database mapping error.  Source database collation ({0}) is not same to target database collation ({1}). Source database: {2}; Target database: {3}.",
```
Comment (AkshayMata) [inline extensions/sql-migration/src/constants/strings.ts:421]: We should add guidance to customer on how to fix as well as a link to a doc

File: extensions/sql-migration/src/wizard/targetSelectionPage.ts [hunk 2]
```diff
@@ -1063,6 +1066,20 @@ export class TargetSelectionPage extends MigrationWizardPage {
 								sourceDatabaseName,
 								mappedSourceDatabaseName));
 					}
+
+					// Collation validation
+					if (sourceDatabaseCollation &&
```
Comment (AkshayMata) [inline extensions/sql-migration/src/wizard/targetSelectionPage.ts:1071]: migrate to private function to do the check, i.e. isCollationSame.

also we should do case insentive compare using toLocaleLowerCase

**Review by AkshayMata — CHANGES_REQUESTED** [review]

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 3]
```diff
@@ -143,6 +143,7 @@ export interface SavedInfo {
 	closedPage: number;
 	databaseAssessment: string[];
 	databaseList: string[];
+	databaseInfoList: Map<string, SourceDatabaseInfo | undefined>;
```
Comment (raymondtruong) [inline extensions/sql-migration/src/models/stateMachine.ts:146]: since you added this `databaseInfoList` in `SavedInfo`, looks like you need to add that here as well:

https://github.com/microsoft/azuredatastudio/blob/7a736b76fa367c3a83c9ee90883373d8da9312ef/extensions/sql-migration/src/dialog/retryMigration/retryMigrationDialog.ts#L37

currently this is causing the build to fail:

image

File: extensions/sql-migration/src/constants/strings.ts (hunk 1 above)
Comment (raymondtruong) [inline extensions/sql-migration/src/constants/strings.ts:421]: I would suggest rewording the string to something like:

"Database mapping error: Source database ({0}) collation ({1}) does not match target database ({2}) collation ({3})."

e.g.

"Database mapping error: Source database (AdventureWorks_source) collation (Latin1_General_CI_AS) does not match target database (AdventureWorks_target) collation (SQL_Latin1_General_CP1_CI_AS)."

Agree with @AkshayMata's comment above to make the message more actionable as well

**Comment by AkshayMata** [comment]
On second thought, would it be better to add the collation check as a step in Brian's new validation dialog: https://github.com/microsoft/azuredatastudio/blob/fd2ced449b8840c6054192e4e9c795d8730340c7/extensions/sql-migration/src/dialog/validationResults/validateIrDialog.ts#L449

The idea here would be to add a new row to "Validate collation"

image

I think this solution could flow better, so just thought I would comment here as food for thought.

A con of this approach is that this dialog is only there for IR code path, so we would need to enable a stripped-down version with just collation check for non-IR code path. On the other hand, standardizing a pattern to validate in 1 place would be helpful especially as it seems like there is a priority to fail fast and earlier within the ADS experience.

**Comment by siyangMicrosoft** [comment]
Will add COLLATION Validation in the validation wizard.

**Comment by Charles-Gagnon** [comment]
@siyangMicrosoft Please fix the commits, did you forget to merge/rebase from main?
